docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,101 @@
+# SDG API :: Add a file format for defining custom Flows
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+block_configs:
+  - block_type: LLMBlock
+    block_config: # LLMBlock constructor kwargs
+      block_name: gen_knowledge
+      output_cols: ["question", "response"]
+    gen_kwargs: # kwargs for block.generate()
+      max_tokens": 2048,
+    drop_duplicates: ["question"]
+  - block_type: FilterByValueBlock
+    block_config:
+      block_name: filter_faithfulness
+      filter_column": judgment
+      filter_value: YES
+      operation: operator.eq
+      batch_kwargs:
+        num_procs: 8
+     drop_columns: ["judgment", "explanation"]
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e.  llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_iters - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+TODO: figure out how these should be supplied to the blocks, and whether any YAML templating based on these are necessary.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
```

Comment:
It would be interesting to extend the config format above to allow for extensions to the `model_family` prompts as well. This could allow users to define prompt templates for other models without needing them hard-coded in the library.

Reply:
Also, it looks like these prompt templates are largely encoding the chat template for the model. For most models that I've seen, the chat template is actually encoded in the model metadata and applied server-side when using the `chat/completions` API. Has there been thought around using the chat API rather than applying the template client-side to the lower-level `completions` API?

Reply:
haven't seen anyone propose it, honestly didn't even realize the difference myself :)

probably outside the scope of this doc, though

Reply:
Yes, definitely outside the scope of this proposal! Just something that would hopefully get simpler in this proposal if it were done more broadly

Reply:
Moving to `chat/completions` would definitely be a separate enhancement, thanks for the suggestion though - I've filed it as #87

Reply:
> It would be interesting to extend the config format above to allow for extensions to the `model_family` prompts as well. This could allow users to define prompt templates for other models without needing them hard-coded in the library.

I guess this would be a "support custom model families" enhancement? Or custom model prompts for known families?

Probably best to file another issue on this to explain the use case.